pkg/server/ca/manager/manager.go

```diff
@@ -550,6 +574,81 @@ func (m *Manager) dropBundleUpdated() {
 	default:
 	}
 }
+
+func (m *Manager) notifyUpstreamAuthoritiesTainted(taintedAuthorities []*x509.Certificate) {
+	select {
+	case m.taintedUpstreamAuthoritiesCh <- taintedAuthorities:
+	default:
+	}
+}
+
+func (m *Manager) fetchRootCAByAuthorityID(ctx context.Context, authorityID string) (*x509.Certificate, error) {
+	bundle, err := m.fetchRequiredBundle(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, rootCA := range bundle.RootCas {
+		if rootCA.TaintedKey {
+			cert, err := x509.ParseCertificate(rootCA.DerBytes)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse RootCA: %w", err)
+			}
+
+			skID := x509util.SubjectKeyIDToString(cert.SubjectKeyId)
+			if authorityID == skID {
+				return cert, nil
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("no tainted root CA found with authority ID: %q", authorityID)
+}
+
+func (m *Manager) processTaintedAuthorities(ctx context.Context, taintedAuthorities []*x509.Certificate) error {
+	// Nothing to rotate if no upstream authority is used
+	if m.upstreamClient == nil {
+		return errors.New("processing of tainted upstream authorities must not be reached when not using upstream authority. Please create an issue.")
+	}
+
+	if len(taintedAuthorities) == 0 {
+		// No tainted keys found
+		return nil
+	}
+
+	m.c.Log.Debug("Processing tainted keys on upstream authority")
+
+	currentSlotCA := m.currentX509CA.x509CA
+	if ok := isX509AuthorityTainted(currentSlotCA, taintedAuthorities); ok {
+		m.c.Log.Debug("Current root CA is signed by a tainted upstream authority, preparing rotation")
```

Comment:
Having the root CA signed by a tainted authority is an unusual event that I think that we should log at Info level.
```suggestion
		m.c.Log.Info("Current root CA is signed by a tainted upstream authority, preparing rotation")
```